syntax = "proto3";

package cloud;

// The AuthService defines all RPCs for user management and file storage
service AuthService {
  // Register a new user (assigns quota)
  rpc Register(RegisterRequest) returns (RegisterResponse);

  // Login with email + password (triggers OTP send)
  rpc Login(LoginRequest) returns (LoginResponse);

  // Verify OTP to confirm authentication
  rpc VerifyOTP(OTPRequest) returns (OTPResponse);

  // Upload a file (authenticated, within quota)
  rpc UploadFile(FileUploadRequest) returns (FileUploadResponse);

  // Download a file
  rpc DownloadFile(FileDownloadRequest) returns (FileDownloadResponse);

  // Delete a file
  rpc DeleteFile(FileDeleteRequest) returns (FileDeleteResponse);

  // List all files stored by a user
  rpc ListFiles(ListFilesRequest) returns (ListFilesResponse);

  // Get quota usage (bytes used vs total)
  rpc GetQuota(QuotaRequest) returns (QuotaResponse);
}

// --- Messages ---

message RegisterRequest {
  string email = 1;
  string username = 2;
  string password = 3;
}

message RegisterResponse {
  bool success = 1;
  string message = 2;
  int64 quota_bytes = 3;   // e.g. 5GB
}

message LoginRequest {
  string email = 1;
  string password = 2;
}

message LoginResponse {
  bool success = 1;
  string message = 2;
}

message OTPRequest {
  string email = 1;
  string otp_code = 2;
}

message OTPResponse {
  bool success = 1;
  string message = 2;
}

// --- File operations ---

message FileUploadRequest {
  string email = 1;
  string filename = 2;
  bytes content = 3;   // actual file bytes
}

message FileUploadResponse {
  bool success = 1;
  string message = 2;
}

message FileDownloadRequest {
  string email = 1;
  string filename = 2;
}

message FileDownloadResponse {
  bytes content = 1;
  string message = 2;
}

message FileDeleteRequest {
  string email = 1;
  string filename = 2;
}

message FileDeleteResponse {
  bool success = 1;
  string message = 2;
}

message ListFilesRequest {
  string email = 1;
}

message FileInfo {
  string filename = 1;
  int64 size = 2;
}

message ListFilesResponse {
  repeated FileInfo files = 1;
}

// --- Quota ---

message QuotaRequest {
  string email = 1;
}

message QuotaResponse {
  int64 used_bytes = 1;
  int64 total_bytes = 2;
}